### PR TITLE
Fix the bug where the localStorage state is invalid.

### PR DESCRIPTION
**Pull request by venkatesh920**

### CODE REVIEW

**Review by IngridMorstrad — CHANGES_REQUESTED** [review]
Please provide a better commit message so the reviewer can identify what was fixed in each of the 2  commits.

File: javascript/photo.js [hunk 1]
```diff
@@ -51,12 +51,12 @@ backgrounds.Photo = new Model({
      * Generates a permutation of size N.
      */
     permutation: function(N) {
-        var numbers = Array();
-        var ind, temp;
+        var i, ind, temp, numbers;
+        numbers = [];
```
Comment (IngridMorstrad) [inline javascript/photo.js:55]: [Question] What's the advantage of this over Array() ?

File: javascript/photo.js [hunk 2]
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
         return imageData;
     },
 
+    isValidLocalStorageState: function(cacheSize) {
+        var count = localStorage.getItem("count");
+        var order = localStorage.getItem("order");
+        var cachedImages = localStorage.getItem("cachedImages");
+        return count != null && Number(count) >= 0 && order != null && JSON.parse(order).length == this.BACKUP_IMAGES_COUNT && cachedImages != null && JSON.parse(cachedImages).length <= cacheSize;
```
Comment (IngridMorstrad) [inline javascript/photo.js]: Is count >= 0 required? Just a sanity check? Similar comments about the other checks.
Also, if possible split this return into three boolean checks for readability if possible.

File: javascript/photo.js [hunk 3]
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
...
-        localStorage.setItem("cacheIndex", 0);
         localStorage.setItem("order", JSON.stringify(order));
         localStorage.setItem("cachedImages", JSON.stringify([]));
     },
 
     /**
      * Displays the image.
-     * TODO: Update function name & doc. This function does more than just display
+     * TODO: Update function name & doc. This function does more than just
+     * display
```
Comment (IngridMorstrad) [inline javascript/photo.js]: Why was this split to 2 lines when the rest of the file exceeds this line length in many places?

File: javascript/photo.js [hunk 4]
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
...
      */
     display: function(numberOfImages, cacheSize) {
-        var count = localStorage.getItem("count");
-        if (count === null) {
+        if (!this.isValidLocalStorageState(cacheSize)) {
```
Comment (IngridMorstrad) [inline javascript/photo.js:143]: Nice!

File: javascript/photo.js [hunk 5]
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
...
         xmlHttp.onreadystatechange = function() {
+            // If the HTTP request is successful, then we store the image received into our cache. If the request is unsuccessful, the cache simply does not get updated.
             if (xmlHttp.readyState == 4 && xmlHttp.status == 200) {
                 var response = JSON.parse(xmlHttp.response);
                 // Fetch an image and cache it so it can be displayed quickly
                 var imageData = parent.parseImage(response, count);
-                var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-                //var cacheIndex = Number(localStorage.getItem("cacheIndex"));
-                cachedImages[cacheIndex] = imageData;
+                cachedImages.push(imageData);
```
Comment (IngridMorstrad) [inline javascript/photo.js:170]: Does this work even though cachedImages is not defined within this function? We had to use parent (this) earlier to solve this.

File: javascript/photo.js [hunk 6]
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
...
-                if (count >= parent.STARTUP_IMAGE_COUNT) {
-                    var imageNumber = count % parent.BACKUP_IMAGES_COUNT;
-                    displayImage(parent.offlineImages[imageNumber]);
-                }
             }
         }
-        // TODO: Think about not calling the API after we get all images (or caching all images & then not calling?)
-        xmlHttp.open("GET", "https://www.reddit.com/r/EarthPorn/top/.json?limit=" + Number(numberOfImages), true);
-        xmlHttp.send(null);
+        // We send an HTTP request only if our cache is not full.
```
Comment (IngridMorstrad) [inline javascript/photo.js:174]: When will it be equal? Assume cacheSize > 1, then each time we add to the cache, on the next call, we will pull from it.

File: javascript/photo.js [hunk 7]
```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
...
-            // TODO: Use variable cacheReadPos & cacheWritePos to prevent errors when fetch fails
-            var cachedImageNumber = cacheIndex - this.STARTUP_IMAGE_COUNT;
-            if (cachedImageNumber < 0) {
-                cachedImageNumber += cacheSize;
-            }
-            var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-            this.displayImage(cachedImages[cachedImageNumber]);
+            this.displayImage(cachedImages[0]);
+            // Once we consume an image from the cache, remove it from the cache.
+            cachedImages.splice(0, 1);
```
Comment (IngridMorstrad) [inline javascript/photo.js:159]: When will the cache size be > this.STARTUP_IMAGE_COUNT? cacheSize variable seems unnecessary.

**Review by IngridMorstrad — CHANGES_REQUESTED** [review]
Nice work overall... :)

File: javascript/photo.js (hunk 6 above)
Comment (venkatesh920) [reply javascript/photo.js:174]: If the startup image count exceeds the cacheSize, then the cache will become full before we can use it.

File: javascript/photo.js (hunk 5 above)
Comment (venkatesh920) [reply javascript/photo.js:170]: cachedImages is in the same scope as parent is.  So it should be fine.

File: javascript/photo.js (hunk 7 above)
Comment (venkatesh920) [reply javascript/photo.js:159]: STARTUP_IMAGE_COUNT and cacheSize shouldn't have any dependency on one another.  For example, STARTUP_IMAGE_COUNT could be equal to BACKUP_IMAGES_COUNT and that shouldn't break how we use the cache.

File: javascript/photo.js (hunk 1 above)
Comment (venkatesh920) [reply javascript/photo.js:55]: No advantage.  They're the same object.  Just seemed like this was more conventional.  Either way is fine - I don't have a strong preference.

**Comment by venkatesh920** [comment]
BTW, we'll squash all the commits into one before merging so don't bother about the individual commits; rather just look at the final diff.

**Review by abhiskk — APPROVED** [review]

File: javascript/photo.js (hunk 5 above)
Comment (IngridMorstrad) [reply javascript/photo.js:170]: Ah... good way of explaining it!

**Review by IngridMorstrad — APPROVED** [review]

**Pull request merged**